File: samples/go/tests/json_schema_kb_nexus_test.go
package tests

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"io"
	"testing"

	"github.com/nexus-rpc/sdk-go/nexus"
	"github.com/stretchr/testify/require"
	"github.com/stretchr/testify/suite"
	"go.temporal.io/sdk/testsuite"
	"go.temporal.io/sdk/workflow"

	"samples/go/kb"
)

// KBNexusIntegrationSuite drives the generated KnowledgeBaseService *service and
// operation definitions* end-to-end through Temporal's in-memory Nexus test
// environment. The caller workflow uses the Temporal SDK's built-in Nexus
// client directly — there is no generated API client — and passes the generated
// operation references (kb.KnowledgeBaseService.GetPage, ...) so the SDK
// type-checks each request/response against the generated definition.
type KBNexusIntegrationSuite struct {
	suite.Suite
	testsuite.WorkflowTestSuite
	env   *testsuite.TestWorkflowEnvironment
	calls []string
}

func (s *KBNexusIntegrationSuite) loadModel(name string, out any) {
	s.Require().NoError(json.Unmarshal(jsonSchemaFixtureBytes(s.T(), "kb", name), out))
}

func (s *KBNexusIntegrationSuite) SetupTest() {
	s.env = s.NewTestWorkflowEnvironment()
	s.calls = nil

	getPage := nexus.NewSyncOperation(kb.KnowledgeBaseService.GetPage.Name(),
		func(_ context.Context, input kb.GetPageInput, _ nexus.StartOperationOptions) (kb.Page, error) {
			s.calls = append(s.calls, "GetPage")
			s.Equal("page-1", input.PageId)
			var page kb.Page
			s.loadModel("page.json", &page)
			return page, nil
		})

	putBlock := nexus.NewSyncOperation(kb.KnowledgeBaseService.PutBlock.Name(),
		func(_ context.Context, input kb.Block, _ nexus.StartOperationOptions) (kb.PutBlockOutput, error) {
			s.calls = append(s.calls, "PutBlock")
			s.Equal("block-1", input.BlockId)
			s.Require().NotNil(input.Style)
			s.Require().NotNil(input.Style.Bold)
			s.True(*input.Style.Bold)
			var out kb.PutBlockOutput
			s.loadModel("put-block-output.json", &out)
			return out, nil
		})

	getCategoryTree := nexus.NewSyncOperation(kb.KnowledgeBaseService.GetCategoryTree.Name(),
		func(_ context.Context, input kb.GetCategoryTreeInput, _ nexus.StartOperationOptions) (kb.Category, error) {
			s.calls = append(s.calls, "GetCategoryTree")
			s.Equal("root", input.RootId)
			var category kb.Category
			s.loadModel("category-tree.json", &category)
			return category, nil
		})

	service := nexus.NewService(kb.KnowledgeBaseService.ServiceName)
	s.Require().NoError(service.Register(getPage, putBlock, getCategoryTree))
	s.env.RegisterNexusService(service)
}

func TestKBNexusIntegrationSuite(t *testing.T) {
	suite.Run(t, &KBNexusIntegrationSuite{})
}

func TestKBNexusInvalidGeneratedInputIsBadRequest(t *testing.T) {
	called := false
	operation := nexus.NewSyncOperation(kb.KnowledgeBaseService.GetPage.Name(),
		func(_ context.Context, _ kb.GetPageInput, _ nexus.StartOperationOptions) (kb.Page, error) {
			called = true
			return kb.Page{}, nil
		})
	service := nexus.NewService(kb.KnowledgeBaseService.ServiceName)
	require.NoError(t, service.Register(operation))
	registry := nexus.NewServiceRegistry()
	require.NoError(t, registry.Register(service))
	handler, err := registry.NewHandler()
	require.NoError(t, err)

	input := nexus.NewLazyValue(nexus.DefaultSerializer(), &nexus.Reader{
		ReadCloser: io.NopCloser(bytes.NewReader([]byte(`{"pageId":null}`))),
		Header:     nexus.Header{"type": "application/json"},
	})
	ctx := nexus.WithHandlerContext(context.Background(), nexus.HandlerInfo{
		Service:   kb.KnowledgeBaseService.ServiceName,
		Operation: kb.KnowledgeBaseService.GetPage.Name(),
	})
	_, err = handler.StartOperation(
		ctx,
		kb.KnowledgeBaseService.ServiceName,
		kb.KnowledgeBaseService.GetPage.Name(),
		input,
		nexus.StartOperationOptions{},
	)
	require.Error(t, err)
	var handlerError *nexus.HandlerError
	require.ErrorAs(t, err, &handlerError)
	require.Equal(t, nexus.HandlerErrorTypeBadRequest, handlerError.Type)
	require.False(t, called, "operation handler must not run for invalid generated input")
}

type kbResult struct {
	BlockId         string
	CategoryChildId string
	PageId          string
	Revision        int64
}

func (s *KBNexusIntegrationSuite) TestOperationsUseRealNexusClient() {
	s.env.ExecuteWorkflow(func(ctx workflow.Context) (kbResult, error) {
		client := workflow.NewNexusClient("knowledge-base", kb.KnowledgeBaseService.ServiceName)

		var page kb.Page
		if err := client.ExecuteOperation(
			ctx, kb.KnowledgeBaseService.GetPage, kb.GetPageInput{PageId: "page-1"},
			workflow.NexusOperationOptions{},
		).Get(ctx, &page); err != nil {
			return kbResult{}, err
		}
		if len(page.Blocks) == 0 {
			return kbResult{}, errors.New("expected page block")
		}

		var putBlockOutput kb.PutBlockOutput
		if err := client.ExecuteOperation(
			ctx, kb.KnowledgeBaseService.PutBlock, page.Blocks[0],
			workflow.NexusOperationOptions{},
		).Get(ctx, &putBlockOutput); err != nil {
			return kbResult{}, err
		}

		var category kb.Category
		if err := client.ExecuteOperation(
			ctx, kb.KnowledgeBaseService.GetCategoryTree, kb.GetCategoryTreeInput{RootId: "root"},
			workflow.NexusOperationOptions{},
		).Get(ctx, &category); err != nil {
			return kbResult{}, err
		}

		var categoryChildID string
		if len(category.Children) > 0 {
			categoryChildID = category.Children[0].Id
		}

		return kbResult{
			BlockId:         putBlockOutput.BlockId,
			CategoryChildId: categoryChildID,
			PageId:          page.PageId,
			Revision:        putBlockOutput.Revision,
		}, nil
	})

	s.True(s.env.IsWorkflowCompleted())
	s.NoError(s.env.GetWorkflowError())

	var result kbResult
	s.NoError(s.env.GetWorkflowResult(&result))
	s.Equal(kbResult{
		BlockId:         "block-1",
		CategoryChildId: "child",
		PageId:          "page-1",
		Revision:        7,
	}, result)
	s.Equal([]string{"GetPage", "PutBlock", "GetCategoryTree"}, s.calls)
}

File: samples/typescript/tests/json-schema-kb-nexus.test.ts
import { fileURLToPath } from "node:url";
import { describe, expect, test } from "vitest";
import * as nexus from "nexus-rpc";

import { chatService } from "../chat/services.ts";
import { knowledgeBaseService } from "../kb/kb/services.ts";
import { blockTransferTypeConverter } from "../kb/content/block/models.ts";
import type { Block } from "../kb/content/block/models.ts";
import { categoryTransferTypeConverter } from "../kb/tree/category/models.ts";
import type { Category } from "../kb/tree/category/models.ts";
import { pageTransferTypeConverter } from "../kb/content/page/models.ts";
import type { Page } from "../kb/content/page/models.ts";
import {
  getCategoryTreeInputTransferTypeConverter,
  getPageInputTransferTypeConverter,
  putBlockOutputTransferTypeConverter,
} from "../kb/kb/models.ts";
import type { PutBlockOutput } from "../kb/kb/models.ts";
import { loadFixture as loadFixtureFrom } from "./json-converter-helper.ts";
import { executeWorkflowWithNexus, withWorkflowEnvironment } from "./helpers.ts";

const workflowsPath = fileURLToPath(
  new URL("./workflows/json-schema-kb.ts", import.meta.url),
);
const wireFixtureDir = new URL("../../wire/json_schema/kb/", import.meta.url);

function loadFixture<T = unknown>(name: string): T {
  return loadFixtureFrom<T>(wireFixtureDir, name);
}

describe("json-schema KB generated Nexus service", () => {
  test("exposes generated service + operation definitions", () => {
    expect(knowledgeBaseService.name).toBe("example.kb.v1.KnowledgeBaseService");
    expect(knowledgeBaseService.operations.getPage.name).toBe("GetPage");
    expect(knowledgeBaseService.operations.putBlock.name).toBe("PutBlock");
    expect(knowledgeBaseService.operations.getCategoryTree.name).toBe(
      "GetCategoryTree",
    );
  });

  // Every operation carries its models' transfer type converters as operation
  // type info, so a protocol integration can apply the conversion without the
  // caller naming the converter. `chatService.operations.ping` (void on both
  // sides) is the counter-case: there is no value to convert, so it carries none.
  test("carries transfer type info on every operation", () => {
    expect(
      knowledgeBaseService.operations.getPage.inputType?.transferTypeConverter,
    ).toBe(getPageInputTransferTypeConverter);
    expect(
      knowledgeBaseService.operations.getPage.outputType?.transferTypeConverter,
    ).toBe(pageTransferTypeConverter);
    expect(
      knowledgeBaseService.operations.putBlock.inputType?.transferTypeConverter,
    ).toBe(blockTransferTypeConverter);
    expect(
      knowledgeBaseService.operations.putBlock.outputType?.transferTypeConverter,
    ).toBe(putBlockOutputTransferTypeConverter);
    expect(
      knowledgeBaseService.operations.getCategoryTree.inputType?.transferTypeConverter,
    ).toBe(getCategoryTreeInputTransferTypeConverter);
    expect(
      knowledgeBaseService.operations.getCategoryTree.outputType?.transferTypeConverter,
    ).toBe(categoryTransferTypeConverter);

    expect(chatService.operations.ping.inputType).toBeUndefined();
    expect(chatService.operations.ping.outputType).toBeUndefined();
  });

  // Register a handler bound to the generated service definition, then run a
  // workflow that calls every operation through the SDK's Nexus client. The SDK
  // applies the generated operation type info at both ends of the Nexus boundary,
  // so application code receives and returns typed models without naming their
  // transfer type converters.
  test("drives every operation through a real Nexus client", async () => {
    await withWorkflowEnvironment(async (env) => {
      const calls: Array<[string, unknown]> = [];
      const handler = nexus.serviceHandler(knowledgeBaseService, {
        async getPage(_ctx, input) {
          calls.push(["GetPage", input]);
          return loadFixture<Page>("page.json");
        },
        async putBlock(_ctx, input) {
          calls.push(["PutBlock", input]);
          return loadFixture<PutBlockOutput>("put-block-output.json");
        },
        async getCategoryTree(_ctx, input) {
          calls.push(["GetCategoryTree", input]);
          return loadFixture<Category>("category-tree.json");
        },
      });

      const result = await executeWorkflowWithNexus<{
        blockId: string;
        categoryChildId: string | undefined;
        pageId: string;
        revision: number;
      }>(env, {
        endpoint: "knowledge-base",
        nexusServices: [handler],
        workflowType: "jsonSchemaKbCaller",
        workflowsPath,
      });

      expect(result).toEqual({
        blockId: "block-1",
        categoryChildId: "child",
        pageId: "page-1",
        revision: 7,
      });

      expect(calls.map(([operation]) => operation)).toEqual([
        "GetPage",
        "PutBlock",
        "GetCategoryTree",
      ]);
      expect((calls[0]?.[1] as { pageId: string }).pageId).toBe("page-1");
      const putBlockInput = calls[1]?.[1] as Block;
      expect(putBlockInput.blockId).toBe("block-1");
      expect(putBlockInput.style?.bold).toBe(true);
      expect((calls[2]?.[1] as { rootId: string }).rootId).toBe("root");
    });
  });
});
